lib/compress/zstd_compress.c

```diff
@@ -202,178 +197,325 @@ size_t ZSTD_sizeof_CStream(const ZSTD_CStream* zcs)
+        }
+        params->dictMode = (ZSTD_dictMode_e)value;
         return 0;
 
-    case ZSTD_p_refDictContent :
-        if (cctx->cdict) return ERROR(stage_wrong);  /* must be set before loading */
-        /* dictionary content will be referenced, instead of copied */
-        cctx->dictContentByRef = value>0;
+    case ZSTD_p_forceMaxWindow :
+        params->forceWindow = value > 0;
         return 0;
 
-    case ZSTD_p_forceMaxWindow :  /* Force back-references to remain < windowSize,
-                                   * even when referencing into Dictionary content
-                                   * default : 0 when using a CDict, 1 when using a Prefix */
-        cctx->forceWindow = value>0;
-        cctx->loadedDictEnd = 0;
-        return 0;
-
-    case ZSTD_p_nbThreads:
-        if (value==0) return 0;
-        DEBUGLOG(5, " setting nbThreads : %u", value);
+    case ZSTD_p_nbThreads :
+        if (value == 0) { return 0; }
 #ifndef ZSTD_MULTITHREAD
         if (value > 1) return ERROR(parameter_unsupported);
 #endif
-        if ((value>1) && (cctx->nbThreads != value)) {
-            if (cctx->staticSize)  /* MT not compatible with static alloc */
-                return ERROR(parameter_unsupported);
-            ZSTDMT_freeCCtx(cctx->mtctx);
-            cctx->nbThreads = 1;
-            cctx->mtctx = ZSTDMT_createCCtx_advanced(value, cctx->customMem);
-            if (cctx->mtctx == NULL) return ERROR(memory_allocation);
-        }
-        cctx->nbThreads = value;
-        return 0;
+        return ZSTDMT_initializeCCtxParameters(params, value);
 
-    case ZSTD_p_jobSize:
-        if (cctx->nbThreads <= 1) return ERROR(parameter_unsupported);
-        assert(cctx->mtctx != NULL);
-        return ZSTDMT_setMTCtxParameter(cctx->mtctx, ZSTDMT_p_sectionSize, value);
+    case ZSTD_p_jobSize :
+        if (params->nbThreads <= 1) return ERROR(parameter_unsupported);
+        return ZSTDMT_CCtxParam_setMTCtxParameter(params, ZSTDMT_p_sectionSize, value);
 
-    case ZSTD_p_overlapSizeLog:
-        DEBUGLOG(5, " setting overlap with nbThreads == %u", cctx->nbThreads);
-        if (cctx->nbThreads <= 1) return ERROR(parameter_unsupported);
-        assert(cctx->mtctx != NULL);
-        return ZSTDMT_setMTCtxParameter(cctx->mtctx, ZSTDMT_p_overlapSectionLog, value);
+    case ZSTD_p_overlapSizeLog :
+        if (params->nbThreads <= 1) return ERROR(parameter_unsupported);
+        return ZSTDMT_CCtxParam_setMTCtxParameter(params, ZSTDMT_p_overlapSectionLog, value);
 
     default: return ERROR(parameter_unsupported);
     }
 }
 
+/**
+ * This function should be updated whenever ZSTD_CCtx_params is updated.
+ * Parameters are copied manually before the dictionary is loaded.
+ * The multithreading parameters jobSize and overlapSizeLog are set only if
+ * nbThreads > 1.
+ *
+ * Pledged srcSize is treated as unknown.
+ */
+size_t ZSTD_CCtx_applyCCtxParams(ZSTD_CCtx* cctx, const ZSTD_CCtx_params* params)
```

Comment:
comment about naming : 
within `CCtx`, we have now `requestedParams` and `appliedParams`.
From an API user perspective, this distinction is invisible.
So `applyCCtxParams` looks fine : just applies parameters from `ZSTD_CCtx_params*` into `ZSTD_CCtx*`. Clear.

From a Zstandard developer perspective though, `cctx->appliedParams` exists, and `ZSTD_CCtx_applyCCtxParams()` seems to refer to it, due to name similarity.
But that's not the case : `ZSTD_CCtx_applyCCtxParams()` only sets parameters inside `cctx->requestedParams`, as it should.

I was wondering if naming could be improved to please both audiences.